### PR TITLE
Diagnostic HUDs

### PR DESCRIPTION
**Pull request by GunHog**
Adds the Diagnostic HUD and NV Diagnostic HUD!
- HUD item that shows health and status of Cyborgs, AI, and Mechs.
- Shows Cyborg health, Alive/Dead/KO status, and cell charge.
- Shows AI status and health.
- Shows mech health, cell charge, and internal damage, if any.
- Integrated to AI to show health of borgs and mechs a Malf would have interest in stealing.
- Integrated to borgs, most useful to Engineering borgs to for repair triage.
- Integrated into RD's hardsuit helmet.

screenshot 2015-10-28 00 38 39
#### Known Issues

~~\- HUDs do not show on newly spawned entities until HUD is taken off and activated again.~~
- Sanity check required to prevent runtime - cause unknown.
- Code may need improvement - Passive HUD code is awful to navigate and modify properly.
- ~~Sprites may need revision. I do not remember who gave them for me for this, however.~~

~~The RD's hardsuit currently has one integrated, but will likely be removed in the final version due to a bug where wearing a diagnostic HUD item with helmet on will cause the HUD to be lost if either the helmet or HUD item is removed.~~

### CODE REVIEW

File: code/modules/mob/living/silicon/silicon.dm [hunk 1]
```diff
@@ -25,9 +27,15 @@
...
 	var/med_hud = DATA_HUD_MEDICAL_ADVANCED //Determines the med hud to use
 	var/sec_hud = DATA_HUD_SECURITY_ADVANCED //Determines the sec hud to use
+	var/d_hud = DATA_HUD_DIAGNOSTIC //There is only one kind of diag hud
 
 	var/law_change_counter = 0
 
+/mob/living/silicon/New()
+	var/datum/atom_hud/data/diagnostic/diag_hud = huds[DATA_HUD_DIAGNOSTIC]
+	diag_hud.add_to_hud(src)
+	..()
```
Comment (RemieRichards) [inline code/modules/mob/living/silicon/silicon.dm]: ..() is called first before any changes in the override, usually.
Just consistency y'know.

**Comment by WJohn** [comment]
:+1: With the sprites, even if they're just orange huds.

File: code/modules/mob/living/silicon/silicon.dm (hunk 1 above)
Comment (GunHog) [inline code/modules/mob/living/silicon/silicon.dm]: Will fix.

**Comment by GunHog** [comment]
I cannot into mapping, so maybe I should add two to the robotics wardrobe locker?

**Comment by Metacide** [comment]
Adding them to the locker sounds good, and they look pretty handy. If you don't add them to the lockers and this gets merged, can you @ me so I remember to add them to MetaStation with the next update?

:+1:

**Comment by tkdrg** [comment]
This fixes #5863

**Comment by GunHog** [comment]
No, it does not. I removed the airlock and machine stuffs. It only works on borgs, AI, and mechs currently. I will add bots eventually once the simple animal thing is added.

Adding airlock status things is even more of a major pain to add properly, and people were stating balance issues with it regarding antags shocking doors.

**Comment by Cookiesndip** [comment]
Very handy for any antag subverting the AI. +

**Comment by Razharas** [comment]
@GunHog 
Tell me when its done

**Comment by GunHog** [comment]
Alrighty! I really do need some sprites and to improve the code a little. I could probably squeeze a few out of WJohn for Xeno code!

**Comment by GunHog** [comment]
New Health Bar sprites and cell power bar! Thanks WJ!

**Comment by GunHog** [comment]
Well, I have not found any *obvious* problems yet. Review when you like, @Razharas!

**Comment by Razharas** [comment]
Didnt notice any if(ckey == "GunHog") so its good enough for me, fix bugs when they are reported

File: code/modules/clothing/spacesuits/hardsuit.dm [hunk 2]
```diff
@@ -318,16 +318,26 @@
...
+		onboard_hud_enabled = 1
+		var/datum/atom_hud/DHUD = huds[DATA_HUD_DIAGNOSTIC]
+		DHUD.add_hud_to(user)
 
-/obj/item/clothing/head/helmet/space/hardsuit/rd/dropped(mob/user)
-	user.scanner.devices -= 1
+/obj/item/clothing/head/helmet/space/hardsuit/rd/dropped(mob/living/carbon/human/user)
 	..(user)
+	user.scanner.devices -= 1
+	if(onboard_hud_enabled && !(user.glasses && istype(user.glasses, /obj/item/clothing/glasses/hud/diagnostic)))
```
Comment (phil235) [inline code/modules/clothing/spacesuits/hardsuit.dm]: this is a hacky fix, and it doesn't even solve the same issue when the items are switched (removing your diag glasses while wearing the helmet).

**Pull request merged**